src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs

```diff
@@ -0,0 +1,218 @@
+            {
+                var analyzers = stateSets.Select(s => s.Analyzer).ToImmutableArrayOrEmpty();
+                return CreateAnalyzerDriverAsync(project, analyzers, includeSuppressedDiagnostics, cancellationToken);
+            }
+
+            public async Task<CompilationWithAnalyzers> CreateAnalyzerDriverAsync(
+                Project project, ImmutableArray<DiagnosticAnalyzer> analyzers, bool includeSuppressedDiagnostics, CancellationToken cancellationToken)
+            {
+                if (!project.SupportsCompilation)
+                {
+                    return null;
+                }
+
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+                // Create driver that holds onto compilation and associated analyzers
+                return CreateAnalyzerDriver(
+                    project, compilation, analyzers, logAnalyzerExecutionTime: false, reportSuppressedDiagnostics: includeSuppressedDiagnostics);
+            }
+
+            private CompilationWithAnalyzers CreateAnalyzerDriver(
+                Project project,
+                Compilation compilation,
+                ImmutableArray<DiagnosticAnalyzer> analyzers,
+                bool logAnalyzerExecutionTime,
+                bool reportSuppressedDiagnostics)
+            {
+                // PERF: there is no analyzers for this compilation.
+                //       compilationWithAnalyzer will throw if it is created with no analyzers which is perf optimization.
+                if (analyzers.IsEmpty)
+                {
+                    return null;
+                }
+
+                Contract.ThrowIfFalse(project.SupportsCompilation);
+                AssertCompilation(project, compilation);
+
+                var analysisOptions = GetAnalyzerOptions(project, logAnalyzerExecutionTime, reportSuppressedDiagnostics);
+
+                // Create driver that holds onto compilation and associated analyzers
+                return compilation.WithAnalyzers(analyzers, analysisOptions);
+            }
+
+            private CompilationWithAnalyzersOptions GetAnalyzerOptions(
+                Project project,
+                bool logAnalyzerExecutionTime,
+                bool reportSuppressedDiagnostics)
+            {
+                // in IDE, we always set concurrentAnalysis == false otherwise, we can get into thread starvation due to
+                // async being used with syncronous blocking concurrency.
+                return new CompilationWithAnalyzersOptions(
+                    options: new WorkspaceAnalyzerOptions(project.AnalyzerOptions, project.Solution.Workspace),
+                    onAnalyzerException: GetOnAnalyzerException(project.Id),
+                    analyzerExceptionFilter: GetAnalyzerExceptionFilter(project),
+                    concurrentAnalysis: false,
```

Comment:
that was my initial implementation and I changed it to project being parallel rather than each compilationWithAnalyzer. until CompilationWithAnalyzer uses async based concurrency rather than synchronous lock based concurrency, it is too dangerous to be used in IDE. it can cause thread starvation which we have hit many times for many different features which tried mix async+lock based concurrency.